docs/design/design.md

```diff
@@ -0,0 +1,131 @@
+- **P2P**: Peer-to-Peer, a distributed application architecture, more info refers to [wiki](https://en.wikipedia.org/wiki/Peer-to-peer).
+- **Task**: A Task will store some meta info about the taskFile, pieces and something else.A Task has a one-to-one correspondence with a file on the disk which is identified by taskID. If you want to download a file from Supernode, you should register a task which means that tell the server about the file info you want to download before actually do it.
+- **DfgetTask**: A DfgetTask represents a download process initiated by Dfget  or other clients. When Dfget try to download a file from p2p network, Supernode will create a DfgetTask object for managing the lifecycle of download process.
+- **Peer**: In a P2P network, peers are both suppliers and consumers of resources. So before Dfget downloads a task from Supernode, Dfget will launch a webserver that provides downloaded files for other peers to download in the P2P network and send a `Peer/Register` request to Supernode to join the P2P network. Only after that can Dfget download files from P2P network.
+- **Piece**: A piece is a part of the file that is going to download, which can be interpreted as a block of files. In dragonfly, the downloading file is not transmitted in a whole file, but in pieces.
+
+## API Design
+
+For a program, a well-designed API is very important, we design the APIs with principles in `RESTFUL` as follows:
+
+- Keep each API's responsibilities simple.
+- Provide APIs that cover the full process life cycle. Now we've just exposed the APIs required by the core process, and we'll add the appropriate APIs as needed later.
+- Use resources as the heart of the API, but they do not have to correspond to an actual data object one-by-one. `E.g.` Tasks  are resources used by a client to operate, DfgetTasks are for internal implementation which should not be exposed to client.
+- URLs should include nouns, not verbs. In additional, use plural nouns instead of singular nouns. `E.g.` `/peers/{id}: GET`.
+- Use HTTP response status codes to represent the outcome of operations on resources. Please refer to [linker](https://github.com/dragonflyoss/Dragonfly/blob/master/apis/swagger.yml#L793).
+- Avoid URL's path level is too deep, we try not to have URLs deeper than resource/identifier/resource, the deepest URL in our APIs is /tasks/{id}/pieces/{pieceRange} because of special circumstances.
+- Well-documented,  we organize the API with swagger and keep the API synchronized with the documentation through swagger generation tools.
+
+### Reference Docs
+
+For API definition, please refer to [swagger.yml](https://github.com/dragonflyoss/Dragonfly/blob/master/apis/swagger.yml)
+
+For API document, please refer to [apis.md](https://github.com/dragonflyoss/Dragonfly/blob/master/docs/apis.md)
+
+## Architecture
+
+In DragonFly, we have two roles:
+
+- Supernode, a SuperNode is a long-time process that plays the following roles:
```

Comment:
To be honest, the name supernode itself is super confusing, LoL.

Reply:
And this section has many duplication comparing with `## Nouns/Phase Explanation`.